File: src/main/java/org/fountainmc/api/entity/hanging/ItemFrame.java
package org.fountainmc.api.entity.hanging;

import org.fountainmc.api.NonnullByDefault;
import org.fountainmc.api.entity.data.EntityData;
import org.fountainmc.api.entity.data.hanging.ItemFrameData;
import org.fountainmc.api.entity.data.hanging.MutableItemFrameData;

/**
 * An ItemFrame that is hanging on a wall
 */
@NonnullByDefault
public interface ItemFrame extends HangingEntity, MutableItemFrameData {
    /**
     * Copy all of the given data to this item frame.
     * <p>Doesn't copy passenger information.</p>
     *
     * @param data the data to copy from
     */
    @Override
    default void copyDataFrom(EntityData data) {
        MutableItemFrameData.super.copyDataFrom(data);
    }

    /**
     * Take a snapshot of this entity's data
     * <p>The resulting snapshot is thread-safe.</p>
     *
     * @return a snapshot
     */
    @Override
    ItemFrameData snapshot();

}

File: src/main/kotlin/org/fountainmc/api/entity/hanging/ItemFrame.kt
package org.fountainmc.api.entity.hanging

import org.fountainmc.api.entity.data.EntityData
import org.fountainmc.api.entity.data.hanging.ItemFrameData

/**
 * An ItemFrame that is hanging on a wall
 */
interface ItemFrame : HangingEntity, ItemFrameData {

    /**
     * Copy all of the mutable properties from the given data into this object.
     *
     * Doesn't copy passenger information.
     * Doesn't copy the [direction] the frame is facing.
     *
     * @param data the data to copy from
     */
    override fun copyDataFrom(data: EntityData) {
        super<ItemFrameData>.copyDataFrom(data)
    }

    /**
     * Take a snapshot of this entity's data
     *
     * The resulting snapshot is thread-safe.
     *
     * @return a snapshot
     */
    override fun snapshot(): ItemFrameData

}
